ocean/helium/data_transfer.py

```diff
@@ -183,13 +201,33 @@ def upload_file(src_path, dest_path, meta):
 
     total_size = sum(f.stat().st_size for f in src_file_list)
 
+    with ThreadPoolExecutor() as executor:
+        # Calculate local ETags in parallel.
+        src_etag_iter = executor.map(_calculate_etag, src_file_list)
+
+        # While waiting for the above, get ETags for the destination dir in the bucket.
+        # (Listing the whole bucket is slow, but the target dir might be a good compromise.)
+        existing_etags = {}
+        for response in _list_object_versions(Bucket=bucket, Prefix=key):
+            for obj in response.get('Versions', []):
+                existing_etags[obj['ETag']] = (obj['Key'], obj['VersionId'])
+
+        src_etag_list = list(src_etag_iter)
+
     with tqdm(total=total_size, unit='B', unit_scale=True) as progress:
         callback = ProgressCallback(progress)
 
         futures = []
-        for f in src_file_list:
+        for f, etag in zip(src_file_list, src_etag_list):
             real_dest_path = key + str(f.relative_to(src_root)) if (not key or key.endswith('/')) else key
-            future = s3_manager.upload(str(f), bucket, real_dest_path, extra_args, [callback])
+            existing_src = existing_etags.get(etag)
+            if existing_src is not None:
+                future = s3_manager.copy(
```

Comment:
throw a one line comment that we are skipping uploads of preexisting files with matching hashes